webapp/store/reducers/nav.js

```diff
@@ -0,0 +1,40 @@
+import { ADD_SIDE_NAV, REMOVE_SIDE_NAV } from '../constants/nav';
+import { initialMetadata } from '../../plugins/plugins';
+import { sortedNavItems } from '../selectors/nav';
+
+const pluginMetadata = initialMetadata();
+const initialState = {
+  sidebar: []
+};
+
+const navStore = (state = initialState, action) => {
+  const newState = { ...state };
+  const { type, payload } = action;
+
+  switch (type) {
+    case ADD_SIDE_NAV: {
+      const sidebar = [...newState.sidebar];
+      sidebar.push(payload);
+      newState.sidebar = sidebar;
+      return newState;
+    }
+
+    case REMOVE_SIDE_NAV: {
+      const sidebar = [...newState.sidebar];
+      const index = sidebar.findIndex(
```

Comment:
`findIndex` will stop the first time its function evaluates to true, is there any case in which it could cause a bug since there is an or clause?

Reply:
So for ones that come directly from plugins we do ensure that there are no duplicate names. However with the sidebar items this is a good point since we only check for duplicates when passing down the state to component props. I'm not sure what the best way to handle this is. One option would be to do the same check we're doing in the selector and save it in the store like that. I'm not sure how I feel about saving something that the user didn't intend to save. We could throw an error or warning in the console and just not save a new path if it's a duplicate, but I'm not thrilled with that either. This way currently basically leaves it up to an implementation detail (the UI developer can add form validation to confirm an input doesn't already exist in the state) but makes sure it doesn't break the UI. This still leaves the removal somewhat ambiguous though as you pointed out. 

Another idea that just occurred to me- maybe we come up with some sort unique id system that gets saved with the nav item? Still wouldn't be trivial to implement because trying to find your id if something has a duplicate name wouldn't be straightforward.

Reply:
If there are two views whose nav objects are both { name: 'my-wallet', path: 'my-wallet' } they would hash to the same thing and we'd have a collision.

Reply:
I'm a little nervous to add things to the state that the user (or in this case the plugin developer) didn't explicitly add or include as I feel like this could result in unexpected results or errors. e.g. When you attempt to delete a sidebar nav you created, you won't be allowed to since you don't have the id (which you didn't create), so you'd have to somehow retrieve/confirm an id matches to your nav item. That said, maybe this would be the opportunity to inform the client that you need to make sure you know the unique id before removing a nav item. 

Also worth noting that this isn't constant after runtime/plugin preprocessing. The nav store is initiated with that but then can be changed by plugins arbitrarily (see the sample gist for an example).

Reply:
maybe add nav actions shouldn't even allow duplicate names, which is the same as the plugin builder with metadata? Then to delete you just pass in a name which we know will be unique and the plugin will be able to keep track of what nav items it has added.

Reply:
Couldn't the sidebar nav id be passed to the dom via a data tag so that it is available for the rest of the application to consume without the user needing to know anything about it?

Reply:
Could the unique path somehow be used to differentiate plugins?

Reply:
It would definitely be simple to not allow two nav items to have the same name and I am not opposed to enforcing that restriction

Reply:
I actually really like the idea dom data-tag idea. The advantage I see with that is that it allows implementers to know exactly which nav item is being targeted given user interactions and not have to be aware about other plugins or nav items. Enforcing a unique name and adding a few chars from a hash as an id on creation seems like a good solution

Reply:
Ok! This functionality should be good to go now. Ended up taking me a down a refactor of dealing with the sidebar nav info that I actually like better too so that was good :). Also used the hashing idea, adding a new utility in bpanel-utils for returning a hash given some arbitrary js object or string.